### PR TITLE
Rewrite badge assignment UI

### PR DESCRIPTION
**Pull request by slomo**
Create a new badge assignment UI

Missing features:

 - it doesn't yet support to set/modify descriptions
 - deleting still sucks

### CODE REVIEW

**Comment by der-john** [comment]
works.

**Comment by der-john** [comment]
Petra and Jana said that the descriptions are highly desired in mein.berlin.de, so I'll see what I can do.

**Comment by slomo** [comment]
I think the old ui allowed to set the "jury statements" if the `showDescription` parameter of the assignment  directive is set to true. I checked the code for and it seems that was never used.

**Comment by slomo** [comment]
In the old ui those "statements" where supported but never used. I recommend that if we need them develop a second ui, which focuses more on those jury statements instead of the addition of bages like the one I added.

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts [hunk 1]
```diff
@@ -26,6 +27,7 @@ export interface IBadge {
     description : string;
     name : string;
     path : string;
```
Comment (der-john) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts]: can we rename this to `assignmentPath` to avoid confusion?

Comment (slomo) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts]: Why not go ahead ;)

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts [hunk 2]
```diff
@@ -118,11 +121,23 @@ export var bindPath = (
             "data", SIBadgeAssignment.nick, "badge"
         ]);
 
-        $q.all(_.map(assignableBadgePaths, (b) => adhHttp.get(b).then(extractBadge))).then((badges) => {
-            scope.badges = badges;
-            var groupPaths : string  = _.union.apply(_, _.map(badges, "groups"));
```
Comment (der-john) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts]: I don't get how something called `groupPaths` could be a string before.

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts [hunk 3]
```diff
@@ -118,11 +121,23 @@ export var bindPath = (
...
+        $q.all(_.map(assignableBadgePaths, (b) => adhHttp.get(b).then(extractBadge))).then((badges : any) => {
+            scope.badges = _.keyBy(badges, "path");
+            var groupPaths : string[]  = _.union.apply(_, _.map(badges, "groups"));
```
Comment (der-john) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts]: what's the difference between this line and the following?
> var groupPaths : string[]  = _.union.(_.map(badges, "groups"));

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts (hunk 2 above)
Comment (xi) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts]: I think it should be `string[]`

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts (hunk 3 above)
Comment (xi) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts]: Related: https://stackoverflow.com/questions/1986896/what-is-the-difference-between-call-and-apply

Summary: `_.union.apply(_, [[1, 2], [1]]) == _.union([1, 2], [1]) != _.union([[1, 2], [1]])`

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts [hunk 4]
```diff
@@ -179,74 +196,44 @@ export var badgeAssignmentCreateDirective = (
...
-            };
+                return adhGetBadges(proposal).then((assignments : IBadge[]) => {
+                    scope.assignments = _.keyBy(assignments, "badgePath");
+                    scope.checkboxes = _.mapValues(scope.assignments, (v) => true);
+                });
+            });
 
             scope.submit = () => {
-                var resource = scope.resource;
-                return adhHttp.put(resource.path, fill(resource, scope, adhCredentials.userPath))
```
Comment (der-john) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts]: `fill` can now be deleted.

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts [hunk 5]
```diff
@@ -179,74 +196,44 @@ export var badgeAssignmentCreateDirective = (
...
-                        if (scope.onSubmit) {
-                            scope.onSubmit();
-                        }
-                    }, (response) => {
-                        scope.serverError = response[0].description;
-                    });
-            };
+                return adhGetBadges(proposal).then((assignments : IBadge[]) => {
+                    scope.assignments = _.keyBy(assignments, "badgePath");
+                    scope.checkboxes = _.mapValues(scope.assignments, (v) => true);
```
Comment (der-john) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts]: I don't understand the second argument yet: all assignments are true? But how do the other checkboxes get created?

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts (hunk 1 above)
Comment (der-john) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts]: I tried... wasn't so easy. :(

**Comment by der-john** [comment]
Very nice! As @slomo pointed out, "descriptions" and deletion still need to be dealt with in other PRs.

**Comment by der-john** [comment]
I have one failing acceptance test that can't be related:
> user password reset error displayed if the email is not associated to an user

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Http/HttpSpec.ts [hunk 6]
```diff
@@ -501,11 +505,18 @@ export var register = () => {
                     expect(request[post2.index].result_path).toBe(post2.path);
                 });
 
+                it("sets the delete flag", () => {
+                    expect(request[delete1.index].body.data[SIMetadata.nick]).toBeDefined();
+                    expect(request[delete1.index].body.data[SIMetadata.nick].deleted).toBeDefined();
+                    expect(request[delete1.index].body.data[SIMetadata.nick].deleted).toBe(true);
+                });
```
Comment (xi) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Http/HttpSpec.ts]: should be in a new `describe` block

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Wrapper.html [hunk 7]
```diff
@@ -1,35 +1,12 @@
...
-            data-show-description="showDescription"
-            data-on-submit="submit"
-            data-on-cancel="cancel"></adh-badge-assignment-create>
+      <adh-badge-assignment-list
+         data-badgeable-path="{{badgeablePath}}"
+         data-pool-path="{{poolPath}}"
+         data-show-description="showDescription"
+         data-on-submit="submit"
+         data-on-cancel="cancel"
+         ></adh-badge-assignment-list>
```
Comment (xi) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Wrapper.html]: Indentation is off

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Wrapper.html [hunk 8]
```diff
@@ -1,35 +1,12 @@
...
     </div>
```
Comment (xi) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Wrapper.html]: It looks like the wrapper directive is no longer useful.

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts [hunk 9]
```diff
@@ -101,6 +103,19 @@ export var bindPath = (
         };
     };
 
+    var collectBadgesByGroup = (groupPaths, badges) => {
+        var badgesByGroup = {};
+        for (var groupPath of groupPaths) {
+            badgesByGroup[groupPath] = [];
+            for (var badge of badges) {
```
Comment (xi) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts]: In #1449 we decided not to use for-of loops. Use `_.forOwn` instead.

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts [hunk 10]
```diff
@@ -101,6 +103,19 @@ export var bindPath = (
...
+                if (_.includes(badge.groups, groupPath)) {
+                    badgesByGroup[groupPath].push(badge.path);
+                }
+            }
+        }
+        return badgesByGroup;
+    };
+
     var getAssignableBadges = (rawOptions) => {
```
Comment (xi) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Badge.ts]: This function should return a promise.

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Http/HttpSpec.ts (hunk 6 above)
Comment (der-john) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Http/HttpSpec.ts]: Why? It is functionality of `withTransaction`, just as the rest of its current `describe` block.

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Wrapper.html (hunk 8 above)
Comment (der-john) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Badge/Wrapper.html]: I agree. I'm taking care of it.

File: src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Http/HttpSpec.ts (hunk 6 above)
Comment (xi) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Http/HttpSpec.ts]: `describe` blocks can be nested, and this one is about the `delete()` method on a transaction, not about `withTransaction` itself.

Comment (der-john) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/Packages/Http/HttpSpec.ts]: ok. thanks.

**Pull request merged**